### PR TITLE
feat: return proper error codes from RPC calls

### PR DESCRIPTION
**Pull request by iljakuklic**
Introduce a mechanism to return the proper error codes from failing RPC calls and use it in the existing RPC methods. This also lays the groundwork for adding arbitrary structured data to RPC errors.

This is to make the future rate limiting errors easier to distinguish from other kinds of errors and to prepare for passing additional machine readable data in the RPC error response.

Related to:
* #219
* #213 

## Breaking changes

* Error codes returned from RPC calls have been changed from generic ones, like -20602 (invalid params) to newly introduced ones that specify the issue more precisely. Anyone relying on these error codes will have to update how they are handled. This is unlikely since the original codes were not particularly informative.
* The error message string has moved from the `data` field of the error object to the `message` field which is a better match. Anyone parsing (not the brightest idea) or reading the error message will have to update which field is read.
* The `data` field has been cleared to make space for error-specific structured data to be returned together with the error code and message.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/agglayer-node/src/rpc/error.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+pub struct RpcError(ErrorObjectOwned);
+
+impl RpcError {
+    /// Create an [RpcError] from a jsonrpsee error object.
+    pub fn form_error_object(err_obj: ErrorObjectOwned) -> Self {
+        Self(err_obj)
+    }
+
+    /// Create a basic [RpcError] with just an error code and a message.
+    pub fn basic(code: i32, message: impl ToString) -> Self {
```
Comment (Freyskeyd) [inline crates/agglayer-node/src/rpc/error.rs:69]:
```suggestion
    pub fn basic<C: Into<i32>>(code: C, message: impl ToString) -> Self {
```

Comment (iljakuklic) [reply crates/agglayer-node/src/rpc/error.rs:69]: I think it's better to keep the interfaces tight, only allowing things that make sense. There is, for example, a `From<bool>` for `i32`. The suggested change would allow `RpcError::basic(false, "foo")`. It's not clear at all that the first argument is some sort of error code.

I'm actually thinking of going in the opposite direction: Introduce a new type for agglayer-specific error codes and accept it here rather than plain `i32`, so that using the named constants is encouraged.

Comment (Freyskeyd) [reply crates/agglayer-node/src/rpc/error.rs:69]: Yes, we can introduce an enum `RpcCode` that resolve to the constant and either impl `From` to use it in the various needed method or have a custom method for it:

```rust
pub const ROLLUP_NOT_REGISTERED: i32 = -10101;

pub enum RpcCode {
  RollupNotRegistered
}

impl From<RpcCode> for i32 {
  fn from(value: RpcCode) -> i32 {
    match value {
      RpcCode::RollupNotRegistered => ROLLUP_NOT_REGISTERED
    }
  }
}

```

Comment (iljakuklic) [reply crates/agglayer-node/src/rpc/error.rs]: So I did something like this but with a struct which is just a wrapper around the code, with a bunch of associated constants. The struct is a bit more open-ended in that it still allows being constructed form arbitrary integers. Over all it probably does not make a massive amount of difference.

File: crates/agglayer-node/src/rpc/error.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    }
+}
+
+#[derive(PartialEq, Eq, Serialize, Debug, thiserror::Error)]
+#[serde(rename_all = "kebab-case")]
+pub enum StatusError {
+    #[error("Transaction not found")]
+    TxNotFound { hash: H256 },
+
+    #[error("Failed get the current L1 block")]
```
Comment (Freyskeyd) [inline crates/agglayer-node/src/rpc/error.rs:69]:
```suggestion
    #[error("Failed to get the current L1 block")]
```

File: crates/agglayer-node/src/rpc/tests.rs [hunk 3]
```diff
@@ -106,6 +107,52 @@ async fn check_tx_status() {
     assert_eq!(res, "done");
 }
 
+#[tokio::test]
+async fn check_tx_status_fail() {
```
Comment (Freyskeyd) [inline crates/agglayer-node/src/rpc/tests.rs:111]: We could add some tests with `insta` to acknowledge the shape of the errors.

File: crates/agglayer-node/src/rpc/error.rs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    #[error("Dry run failed")]
+    DryRun { detail: String },
+
+    #[error("State root verification failed")]
+    RootVerification { detail: String },
+}
+
+#[derive(PartialEq, Eq, Serialize, Debug, thiserror::Error)]
+#[serde(rename_all = "kebab-case")]
+pub enum SettlementError {
```
[the remark this answers is not in the document]
Comment (iljakuklic) [reply crates/agglayer-node/src/rpc/error.rs:39]: This is mostly due to the `Serialize` instance which is used to populate the `"data"` field of the json error. Putting `Serialize` on `kernel::SettlementError` would probably require a custom impl or heavy annotation since most of the fields are not `Serialize`. Another way would be to come up with a different way of creating the structured data but I find the `Serialize`-based approach rather neat.

Also the error messages provided by thiserror here are a bit more brief. They end up in the `"message"` field, leaving most of the detail to the `"data"` field.

In summary this approach gives us control over how the json error is rendered without compromising on information stored in `kernel::SettlementError`.

File: crates/agglayer-node/src/rpc/tests.rs (hunk 3 above)
Comment (iljakuklic) [reply crates/agglayer-node/src/rpc/tests.rs:113]: Added a bunch of them covering all the error cases defined in the `rpc::error` module.

File: crates/agglayer-node/src/rpc/error.rs (hunk 1 above)
Comment (iljakuklic) [reply crates/agglayer-node/src/rpc/error.rs]: Resolving this thread as the discussion does not apply anymore
